consensus/fork.go

```diff
@@ -4,119 +4,115 @@ import (
-// how they were before the call was made.
+// forkBlockchain will move the consensus set onto the 'newNode' fork. An error
+// will be returned if any of the blocks applied in the transition are found to
+// be invalid. forkBlockchain is atomic; the State is only updated if the
+// function returns nil.
 func (s *State) forkBlockchain(newNode *blockNode) (err error) {
-	// Get the state hash before attempting a fork.
-	var stateHash crypto.Hash
+	// In debug mode, record the old state hash before attempting the fork.
+	// This variable is otherwise unused.
+	var oldHash crypto.Hash
 	if DEBUG {
-		stateHash = s.stateHash()
+		oldHash = s.Hash()
 	}
 
-	// Get the list of blocks tracing from the new node to the blockchain, then
-	// rewind to the common parent.
-	backtrackNodes := s.backtrackToBlockchain(newNode)
-	commonParent := backtrackNodes[len(backtrackNodes)-1]
-	rewoundNodes := s.rewindToNode(commonParent)
+	// determine common parent
+	commonParent := s.backtrackToBlockchain(newNode)[0]
 
-	// Update the consensus to include all of the block nodes that go from the
-	// common parent to `newNode`. If any of the blocks are invalid, reverse
-	// all of the changes and switch back to the original block.
-	var appliedNodes []*blockNode
-	for i := len(backtrackNodes) - 2; i >= 0; i-- {
-		bn := backtrackNodes[i]
-		appliedNodes = append(appliedNodes, bn)
+	// save currentBlockNode in case something goes wrong
+	oldHead := s.currentBlockNode()
 
-		// If the diffs for this node have already been generated, apply diffs
-		// directly instead of generating them. This is much faster.
-		if bn.diffsGenerated {
-			direction := true // the blockNode is being applied, direction is set to true.
-			s.applyDiffSet(bn, direction)
-			continue
-		}
+	// rewind to the common parent
+	s.rewindToNode(commonParent)
 
-		// If the diffs have not been generated, call generateAndApplyDiff.
-		// This call will fail if the block is somehow invalid. If the call
-		// fails, all of the applied blocks will be reversed, and all of the
-		// rewound blocks will be reapplied, restoring the consensus of the
-		// State to its original condition.
-		err = s.generateAndApplyDiff(bn)
-		if err != nil {
-			// Mark the invalid block, then rewind all the new blocks and
-			// reapply all of the rewound blocks.
-			s.invalidateNode(bn)
-			s.rewindToNode(commonParent)
-			for i := len(rewoundNodes) - 1; i >= 0; i-- {
-				direction := true // the blockNode is being applied, direction is set to true.
-				s.applyDiffSet(rewoundNodes[i], direction)
+	// fast-forward to newNode
+	err = s.applyUntilNode(newNode)
+	if err != nil {
+		// restore old path
+		s.rewindToNode(commonParent)
+		errReapply := s.applyUntilNode(oldHead)
+		if DEBUG {
+			if errReapply != nil {
+				panic("couldn't reapply previously applied diffs!")
+			} else if s.Hash() != oldHash {
+				panic("state hash changed after an unsuccessful fork attempt")
```

Comment:
ugh I know, it's pretty gross.
`errReapply` is actually what was causing my forking code to fail tests. I was reusing the `err` variable, so it would get cleared to `nil` before being returned, even if there actually was an error!